### ISSUE TITLE
Circular objects overlap

### ISSUE THREAD

**Issue by gaminn** [erincatto/Box2D#425]
I create circular objects in Box2D:
```
b2BodyDef bd;
bd.position = b2Vec2(1.0, 1.0);
bd.type = b2_dynamicBody;

m_body = world->CreateBody(&bd);

b2CircleShape shape;
shape.m_radius = 0.25; // can vary 0.1 - 0.5

b2FixtureDef fd;
fd.shape = &shape;
m_body->CreateFixture(&fd);
```
I move them using
`m_body->ApplyForce()`

I redraw world via:
```
float32 timeStep = 1/(float)40;//the length of time passed to simulate (seconds)
int32 velocityIterations = 8; //how strongly to correct velocity
int32 positionIterations = 3; //how strongly to correct position

m_world->Step( timeStep, velocityIterations, positionIterations);
```
However, when the objects get in proximity, they overlap. Why and how to fix it? When I change the shape of the objects to rectangle, then everything is OK, objects don't overlap when they get in proximity.
![circle](https://cloud.githubusercontent.com/assets/20270888/19946649/13a49b60-a145-11e6-921b-aedee6969bca.png)

**Comment by louis-langholtz** [comment]
@gaminn Please note that I'm just an interested party with my own Box2D fork that I work with; not the original developer of Box2D nor a developer with any kind of contribution authority to Erin's codebase. With that said, thank you for taking the time to provide this issue report. I'd like to help.

I'm not aware of a bug in Box2D's handling of circles that would have the effect you've described. If you place circle bodies exactly on top of each other however (exactly overlapping), Box2D's Step method won't reposition them short of influences (like separating impulses). Meanwhile Box2D's Step method will reposition polygon-shaped bodies.

You might rule this out as a possible cause by setting different initial positions for each body such that no two bodies overlap to begin with and see if you still have a problem. I recall Box2D's documentation advising using different initial positions, so ensuring that may be wise even if it doesn't fix the problem you're seeing.

If that doesn't fix the problem, I'll look into the problem more if you provide more details. Thanks again.

**Comment by gaminn** [comment]
Thanks for the help. I have to add some information to my situation. The white circle was created with b2_dynamicBody flag and sits in the centre of the screen. The rest of the circles are dynamic bodies, created one by one every 5 seconds at the bottom of the screen (when they are created they don't overlap) and moved by m_body->ApplyForce() towards the white circle. The movement is slow.

**Comment by louis-langholtz** [comment]
@gaminn Thanks for the update!

Some questions:

1. What's gravity set to?
2. What are the fixtures' density and restitution values set to?
3. Do the circle bodies that overlap eventually move apart again?
4. Is there space available along the axis defined by the centers of overlapping circles?

Box2D will allow bodies to overlap some but if they overlap too much Box2D will reposition them to resolve the excess overlap. This threshold is related to the `b2_linearSlop` setting. At radiuses of .1 to .5 though I wouldn't expect to see as much overlap unless this setting was changed from its default or there's no more space for the bodies to be moved away to. Like if they're being compressed exactly inline with each other and the bodies are circles.

Circle repositioning is done along the axis defined by the delta of their positions. Rectangle repositioning meanwhile is done differently such that if all objects were stacked vertically, some rectangles would see some horizontal displacement as well. I could see this being the problem and being the difference you've described if the bodies are being compressed exactly inline with each other.

Is that what the problem is?

**Comment by gaminn** [comment]
1. Gravity is zero.
2. Tried both 1.0 and 0.1. Same effect.
3. Yes, they can move apart again if I apply negative force to them.
4. Yes, The overlapping can occur when there are few (two, three...) circles.

I set b2_linearSlop = 0.0005

I modified the axis to be horizontal axis (circles are generated on the left side of the screen and they move towards white circle) and on this axis, circles do not overlap. I really don't understand this.
![circle2](https://cloud.githubusercontent.com/assets/20270888/19950590/af6b8eec-a158-11e6-8b57-ded15384025c.png)
